src/rebase.js

```diff
@@ -116,13 +116,22 @@ module.exports = (function(){
   function _fetch(endpoint, options){
     _validateEndpoint(endpoint);
     optionValidators.context(options);
-    optionValidators.then(options);
     options.queries && optionValidators.query(options);
     var ref = firebase.database().ref(endpoint);
     ref = _addQueries(ref, options.queries);
-    ref.once('value', (snapshot) => {
+    return ref.once('value').then(snapshot => {
       var data = options.asArray === true ? _toArray(snapshot) : snapshot.val();
-      options.then.call(options.context, data);
+      if(options.then){
+        options.then.call(options.context, data);
+      }
+      return data;
+    }, err => {
+      //call onFailure callback if it exists otherwise return a rejected promise
+      if(options.onFailure && typeof options.onFailure === 'function'){
+        options.onFailure.call(options.context, err);
+      } else {
```

Comment:
> I sure hope not, but I will definitely double check tonight.

Please do. And if they _do_ produce similar errors (which is what I'm sure they do), I think it's OK to leave them in `fetch` as well for consistency, but maybe bump the version to mark breaking changes, and maybe mention some workarounds (e.g. `window.addEventListener('unhandledrejection', event => ...)`, etc.) in documentation. On the other hand, if they _don't_, that'll mean that underlying Firebase API is not as consistent as my perfectionism wants it to be (and it is already the case: somehow only `fetch` method supports separate failure callback now), and your current solution will do just fine in real world.

> That is what I was thinking of trying in the comment on line 123 below.

Yeah, I see